### ISSUE TITLE
allow deactivation of engine only for specific category

### ISSUE THREAD

**Issue by pointhi** [searx/searx#205]
currently, a engine is fully deactivated, if you set it "disabled" only in one of the categories.

**Problems**
1. This isn't intuitive to the users (Probably the user only want the engine deactivated in this specific topic)
2. If the user doesn't activate the engine in all categories, he cannot turn it on again.

**Further Idea**

I think it is clear that searx is not searching inside search-engines, which are deactivated. But allowing the explicite avtivation of a deactivated engine via !bang would be a good idea, and would't hurt anyone.

**Comment by dalf** [comment]
I agree.
(more or less related to #109 )

**Comment by Cqoicebordel** [comment]
:+1: for both.

We need more granularity of config in the user side of things, and also, in the admin side of it.
Having to modify a config file and restart the server isn't a good thing. We should have an admin interface.

**Comment by pointhi** [comment]
I think we should add a database which can handle all of this database-things inside searx. With that, we could also fix #199.

Probably something high-level which can use SQLite as well as MySQL and PostgreSQL, to be capable for later application which could require more speed/efficient, and to be more flexible for admin-side.

**Comment by dalf** [comment]
DB-API is only for relational database :
https://wiki.python.org/moin/DbApiModuleComparison

Other framework are may be too much ?
http://docs.python-guide.org/en/latest/scenarios/db/

**Comment by Cqoicebordel** [comment]
I think something like SQLite would suffice, even for future use.

**Comment by pointhi** [comment]
I think the [peewee](http://docs.peewee-orm.com/en/latest/) framework would be a good choise.

Lightweight, and support of SQLite, MySQL and Postgres by default.

**Comment by dalf** [comment]
I don't know how peewee handle multiprocess : https://github.com/coleifer/peewee/issues/263

**Comment by pointhi** [comment]
I think that's more an issue for the underlying database, because that has to handle multible queries. Peewee is what I'm know only a high-level interface to the database. 

So multible Peewee processes shouldn't interference, as long as the database is handling the queries correctly. MySQL and Postges shouldn't be problematic because there are standalone.

SQLite otherwise could create problems, if we spawn multible instances of SQLite.

**Comment by dalf** [comment]
And as long there is no cache on the Peewee layer.

**Comment by Cqoicebordel** [comment]
@pointhi > For SQLite, only if we write in those different instance. And still, if we allow a little bit of wait on the write requests, it shouldn't be an issue. SQLite is very very good at this :)
The advantages of SQLite are zeroconf, and it is serverless. And you don't need a lib to use it in python.

EDIT : Just so you know, the kind of code I'm using with SQLite in another project : 

``` python
import sqlite3 as lite
con = lite.connect('base.sqlite')
con.text_factory = str
sql = "INSERT INTO feed VALUES(?,?)"
params = ("Text1", "Text2")
with con:
    cur = con.cursor()
    try:
        cur.execute(sql, params);
    except lite.IntegrityError:
        pass
```

It's that easy...

**Comment by pointhi** [comment]
Ok, I have found out that SQLite3 is designed to work with multiple processes. So It shouldn't be a problem for SQLite as long as we are using SQLite3.

It look like Peewee is capable for caching. If we can turn it off or it doesn't affect our querys, I would prefere Peewee instead of only SQLite, because we are more flexible with it.

**Comment by Cqoicebordel** [comment]
Sound good to me.
(even if flexibility mean also dependency with peewee)

(the only issue I ran into was using PHP, who set a default busy timeout of 0, so when SQLite was busy, I got an error. I set it to a value of 2 sec, and was good to go)

**Comment by pointhi** [comment]
I have created a wiki-site for mindstorming: [database-structure](https://github.com/asciimoo/searx/wiki/database-structure)

Various tables are clear, how to build them. But other are very tricky, to be simple as well as flexibel.

**Comment by asciimoo** [comment]
i'm not sure, that this issue really requires a database.

**Comment by pointhi** [comment]
I'm sure solving this issue don't require a database.

Probably we should move this discussion to a new issue, because adding a database is at least necessary for #199, and could also be used to improve the speed of https-rewriting, caching images on server-side, adding an admin-panel,...

**Comment by Cqoicebordel** [comment]
Exactly. It's not for this issue specifically, but a small base is needed to do some things properly, like actualising the conf, the icons, the themes... without restarting the server.
And we can add an admin interface, for the user and the server admin, that has more granularity than now, like allowing an engine for a category but not another.
